### PR TITLE
*: check the schema-validity when the DDL fails

### PR DESCRIPTION
**Pull request by zimulala**
Thank you for working on TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

## What have you changed? (mandatory)

The schema in the file of  ddl_api.go may not be the latest. 
- If we can't pass the check in ddl_api.go, it will return an error. The reason for this error may be due to the schema version is outdated. So we need to distinguish the error. If the schema is out of date, we will return the error of "ErrInfoSchemaExpired" or "ErrInfoSchemaChanged".
- If we can pass the check in ddl_api.go, we will check again in ddl_worker.go with the latest schema.

Move the structure of schemaLeaseChecker to util/schemachecker/schema_checker.go.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested (mandatory)?

unit tests.

## Does this PR affect documentation (docs/docs-cn) update? (optional)

No.

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ package ddl
 import (
 	"fmt"
 	"sync"
+	//"sync/atomic"
```
Comment (shenli) [inline ddl/ddl.go]: Remove this.

File: sessionctx/stmtctx/stmtctx.go [hunk 2]
```diff
@@ -27,6 +27,8 @@ import (
 type StatementContext struct {
 	// Set the following variables before execution
 
+	// IsDDLJobDone is used to mark whether to enter the function of 'doDDLJob'.
+	IsDDLJobDone         bool
```
[the remark this answers is not in the document]
Comment (zimulala) [reply sessionctx/stmtctx/stmtctx.go]: No, it means the SQL transform into the DDL job, and it is done.

File: ddl/ddl.go [hunk 3]
```diff
@@ -396,9 +396,21 @@ func (d *ddl) GetLease() time.Duration {
...
 
-// GetInformationSchema get the infoschema binding to d. It's expoted for testing.
-func (d *ddl) GetInformationSchema() infoschema.InfoSchema {
-	return d.infoHandle.Get()
+// IsInTest is used to determine whether it is in the test.
+var IsInTest bool
+
+// GetInformationSchema gets the infoschema binding to d. It's expoted for testing.
+func (d *ddl) GetInformationSchema(ctx sessionctx.Context) infoschema.InfoSchema {
+	if !IsInTest {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl.go]: Now it is only updated in TestParallelDDLBeforeRunDDLJob, and it's serial.
Maybe I can change the way.

**Comment by zimulala** [comment]
PTAL @winkyao @shenli @coocood

File: executor/ddl.go [hunk 4]
```diff
@@ -38,12 +39,27 @@ type DDLExec struct {
 	done bool
 }
 
+func (e *DDLExec) toErr(err error) error {
+	if e.ctx.GetSessionVars().StmtCtx.IsDDLJobDone {
+		return err
+	}
+
```
Comment (winkyao) [inline executor/ddl.go]: Add some description here.

File: ddl/ddl_db_change_test.go [hunk 5]
```diff
@@ -628,3 +631,97 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
...
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		return info
+	}
+	d := s.dom.DDL()
+	d.SetHook(callback)
+
+	// Make sure we
```
Comment (winkyao) [inline ddl/ddl_db_change_test.go]: make sure we?

File: ddl/ddl_db_change_test.go [hunk 6]
```diff
@@ -628,3 +631,97 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
...
+	ch := make(chan struct{})
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+
+		close(ch)
```
Comment (winkyao) [inline ddl/ddl_db_change_test.go]: Why need a channel here?

Comment (zimulala) [reply ddl/ddl_db_change_test.go]: Hope to run this goroutine before running the next goroutine.

File: ddl/callback.go [hunk 7]
```diff
@@ -28,6 +33,8 @@ type Callback interface {
 	OnJobUpdated(job *model.Job)
 	// OnWatched is called after watching owner is completed.
 	OnWatched(ctx context.Context)
+	// OnGetInfoSchema is called in getting information schema. It's used for testing now.
+	OnGetInfoSchema(ctx sessionctx.Context, fn GetInfoSchema) infoschema.InfoSchema
```
Comment (shenli) [inline ddl/callback.go]: OnGetInfoSchema is a hook which is called in the function ddl.GetInfoSchema(). It is used in the tests.

File: ddl/ddl.go [hunk 8]
```diff
@@ -396,9 +396,21 @@ func (d *ddl) GetLease() time.Duration {
 	return lease
 }
 
-// GetInformationSchema get the infoschema binding to d. It's expoted for testing.
-func (d *ddl) GetInformationSchema() infoschema.InfoSchema {
-	return d.infoHandle.Get()
+// IsInTest is used to determine whether it is in the test.
+var IsInTest bool
```
Comment (shenli) [inline ddl/ddl.go]: Can we remve it?

File: domain/schema_validator.go [hunk 9]
```diff
@@ -173,6 +173,12 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
 
 	// Schema changed, result decided by whether related tables change.
 	if schemaVer < s.latestSchemaVer {
+		// The DDL relatedTableIDs is empty.
+		if len(relatedTableIDs) == 0 {
+			log.Infof("[domain-ddl] the related table ID is empty, checking %d", schemaVer)
+			return ResultFail
```
Comment (shenli) [inline domain/schema_validator.go]: Why return fail when the DDL relatedTableIDs is empry?

File: executor/ddl.go [hunk 10]
```diff
@@ -38,12 +39,27 @@ type DDLExec struct {
 	done bool
 }
 
+func (e *DDLExec) toErr(err error) error {
```
Comment (shenli) [inline executor/ddl.go]: What is this function used for?

File: util/schemachecker/schema_checker.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+// SchemaLeaseChecker is used for checking schema-validity.
+type SchemaLeaseChecker struct {
+	domain.SchemaValidator
+	schemaVer       int64
+	relatedTableIDs []int64
+}
+
+var (
+	// SchemaOutOfDateRetryInterval is the sleeping time when we fail to try.
```
Comment (shenli) [inline util/schemachecker/schema_checker.go]: s/sleeping time/backoff time/

Comment (shenli) [reply util/schemachecker/schema_checker.go]: "fail to try" or "try"

File: util/schemachecker/schema_checker.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+	SchemaOutOfDateRetryInterval = int64(500 * time.Millisecond)
+	// SchemaOutOfDateRetryTimes is upper bound of retry times when the schema is out of date.
```
Comment (shenli) [inline util/schemachecker/schema_checker.go]: s/upper bound of retry times/max retry count/

File: util/schemachecker/schema_checker.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+	SchemaOutOfDateRetryTimes = int32(10)
+)
+
+// NewSchemaChecker creates a new schema checker.
```
Comment (shenli) [inline util/schemachecker/schema_checker.go]: s/NewSchemaChecker/NewSchemaLeaseChecker/

File: util/schemachecker/schema_checker.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/metrics"
+)
+
+// SchemaLeaseChecker is used for checking schema-validity.
+type SchemaLeaseChecker struct {
```
Comment (shenli) [inline util/schemachecker/schema_checker.go]: Add more comments about how the checker works.

File: util/schemachecker/schema_checker_test.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+func TestT(t *testing.T) {
+	CustomVerboseFlag = true
+	TestingT(t)
+}
+
+var _ = Suite(&testSuite{})
+
+type testSuite struct{}
```
Comment (shenli) [inline util/schemachecker/schema_checker_test.go]: Add leak test in SetUp and TearDown.

File: util/schemachecker/schema_checker.go (hunk 11 above)
Comment (zimulala) [reply util/schemachecker/schema_checker.go]: It's original comments.
Do you have a suggestion?

File: domain/schema_validator.go (hunk 9 above)
Comment (zimulala) [reply domain/schema_validator.go]: The DDL relatedTableIDs is always empty. If `schemaVer < s.latestSchemaVer` is true when executing DDL，we return `ResultFail`.
The DDL statements get relatedTableIDs is complicated, so I didn't do it.

File: executor/ddl.go [hunk 16]
```diff
@@ -38,12 +39,30 @@ type DDLExec struct {
...
 }
 
+// toErr converts the error to the ErrInfoSchemaChanged when the schema is outdated.
+func (e *DDLExec) toErr(err error) error {
+	if e.ctx.GetSessionVars().StmtCtx.IsDDLJobReady {
+		return err
+	}
+
+	// Before the DDL job is ready，it encouters an error that may be due to the outdated schema information.
+	// Here we distinguish the ErrInfoSchemaChanged error from other errors.
```
Comment (winkyao) [inline executor/ddl.go]: Add some description to explain why we don't have to check whether the schema is outdated when `e.ctx.GetSessionVars().StmtCtx.IsDDLJobReady` is true.

**Comment by ciscoxll** [comment]
@zimulala Please resolve code conflicts.

**Comment by zimulala** [comment]
PTAL @ciscoxll @shenli

File: ddl/ddl.go [hunk 17]
```diff
@@ -383,9 +383,14 @@ func (d *ddl) GetLease() time.Duration {
 	return lease
 }
 
-// GetInformationSchema get the infoschema binding to d. It's expoted for testing.
-func (d *ddl) GetInformationSchema() infoschema.InfoSchema {
-	return d.infoHandle.Get()
```
Comment (ciscoxll) [inline ddl/ddl.go]: Some DDL actions are in use `d.infoHandle.Get()`, no use `GetInformationSchema` will have what problem?
#### For example
* `AddIndex`
* `AddColumn`
* `DropColumn`

Comment (zimulala) [reply ddl/ddl.go]: @ciscoxll 
No.
In "GetInformationSchema" I add the hook that uses to test. So it's OK.

Comment (ciscoxll) [reply ddl/ddl.go]: Got it.

**Comment by zimulala** [comment]
PTAL @tiancaiamao @coocood

File: ddl/callback.go [hunk 18]
```diff
@@ -28,6 +33,8 @@ type Callback interface {
 	OnJobUpdated(job *model.Job)
 	// OnWatched is called after watching owner is completed.
 	OnWatched(ctx context.Context)
+	// OnGetInfoSchema is a hook which is called in the function ddl.GetInfoSchema(). It is used in the tests.
+	OnGetInfoSchema(ctx sessionctx.Context, fn GetInfoSchema) infoschema.InfoSchema
```
Comment (coocood) [inline ddl/callback.go]: I think this function doesn't belong to the callback.
It's actually an intercepter.

**Comment by zimulala** [comment]
PTAL @coocood

File: ddl/callback.go [hunk 19]
```diff
@@ -14,11 +14,30 @@
...
 )
 
-// Callback is the interface supporting callback function when DDL changed.
+// GetInfoSchema is used for getting information schema in testing.
+type GetInfoSchema func() infoschema.InfoSchema
+
+// Intercept is used for DDL.
+type Intercept interface {
+	// OnGetInfoSchema is an intercept which is called in the function ddl.GetInfoSchema(). It is used in the tests.
+	OnGetInfoSchema(ctx sessionctx.Context, fn GetInfoSchema) infoschema.InfoSchema
```
Comment (coocood) [inline ddl/callback.go]: I think replacing `fn GetInfoSchema` to just `infoschema.InfoSchema` is better.

**Comment by zimulala** [comment]
PTAL @coocood

File: executor/ddl.go [hunk 20]
```diff
@@ -38,12 +39,32 @@ type DDLExec struct {
 	done bool
 }
 
+// toErr converts the error to the ErrInfoSchemaChanged when the schema is outdated.
+func (e *DDLExec) toErr(err error) error {
+	if e.ctx.GetSessionVars().StmtCtx.IsDDLJobReady {
+		return err
```
Comment (tiancaiamao) [inline executor/ddl.go]: `errors.Trace(err)`

File: executor/ddl.go [hunk 21]
```diff
@@ -38,12 +39,32 @@ type DDLExec struct {
...
+	}
+
+	// Before the DDL job is ready, it encouters an error that may be due to the outdated schema information.
+	// After the DDL job is ready, the ErrInfoSchemaChanged error won't happen because we are getting the schema directly from storage.
+	// So we needn't to consider this condition.
+	// Here we distinguish the ErrInfoSchemaChanged error from other errors.
+	dom := domain.GetDomain(e.ctx)
+	checker := schemachecker.NewSchemaChecker(dom, e.is.SchemaMetaVersion(), nil)
+	schemaInfoErr := checker.Check(e.ctx.Txn().StartTS())
```
Comment (tiancaiamao) [inline executor/ddl.go]: The check result won't be `ErrInfoSchemaChanged`, because tableIDs is nil.

File: executor/ddl.go [hunk 22]
```diff
@@ -66,7 +87,7 @@ func (e *DDLExec) Next(ctx context.Context, chk *chunk.Chunk) (err error) {
 		err = e.executeRenameTable(x)
 	}
 	if err != nil {
```
Comment (tiancaiamao) [inline executor/ddl.go]: If err == nil, there is no check for schema outdated?

File: ddl/ddl_db_change_test.go [hunk 23]
```diff
@@ -629,3 +632,99 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
...
+
+	intercept := &ddl.TestIntercept{}
+	var (
+		testSessionCnt        int32
+		testObtainedSchemaCnt int32 // testObtainedSchemaCnt is the number of sessions that have obtained information schema.
+		testFirstConnID       = uint64(math.MaxUint64)
+	)
+	intercept.OnGetInfoSchemaExported = func(ctx sessionctx.Context, is infoschema.InfoSchema) infoschema.InfoSchema {
+		// The following code is for testing.
+		// Make srue the two sessions get the same information schema before executing DDL.
```
Comment (tiancaiamao) [inline ddl/ddl_db_change_test.go]: s/srue/sure

File: domain/schema_validator.go [hunk 24]
```diff
@@ -173,6 +173,13 @@ func (s *schemaValidator) Check(txnTS uint64, schemaVer int64, relatedTableIDs [
 
 	// Schema changed, result decided by whether related tables change.
 	if schemaVer < s.latestSchemaVer {
+		// The DDL relatedTableIDs is empty.
+		if len(relatedTableIDs) == 0 {
```
[the remark this answers is not in the document]
Comment (zimulala) [reply domain/schema_validator.go]: The DDL relatedTableIDs is empty.

File: server/server_test.go [hunk 25]
```diff
@@ -537,21 +537,22 @@ func runTestErrorCode(c *C) {
 		_, err = txn2.Exec("select * from tbl_not_exists;")
 		checkErrorCode(c, err, tmysql.ErrNoSuchTable)
 		_, err = txn2.Exec("create database test;")
-		checkErrorCode(c, err, tmysql.ErrDBCreateExists)
+		// Make tests stable. Some times the error may be the ErrInfoSchemaChanged.
+		checkErrorCode(c, err, tmysql.ErrDBCreateExists, tmysql.ErrUnknown)
 		_, err = txn2.Exec("create database aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa;")
-		checkErrorCode(c, err, tmysql.ErrTooLongIdent)
+		checkErrorCode(c, err, tmysql.ErrTooLongIdent, tmysql.ErrUnknown)
```
[the remark this answers is not in the document]
Comment (zimulala) [reply server/server_test.go]: The error is ErrInfoSchemaChanged.

File: executor/ddl.go (hunk 21 above)
Comment (zimulala) [reply executor/ddl.go]: If "ResultFail" is true, the result is ErrInfoSchemaChanged.

File: executor/ddl.go (hunk 22 above)
Comment (zimulala) [reply executor/ddl.go]: Yes, it will be check in ddl_worker.go

File: server/server_test.go [hunk 26]
```diff
@@ -573,10 +574,19 @@ func runTestErrorCode(c *C) {
...
 	me, ok := e.(*mysql.MySQLError)
 	c.Assert(ok, IsTrue, Commentf("err: %v", e))
-	c.Assert(me.Number, Equals, code)
+	if len(codes) == 1 {
+		c.Assert(me.Number, Equals, codes[0])
+	}
+	isMatchCode := false
+	for _, code := range codes {
+		if me.Number == code {
+			isMatchCode = true
```
[the remark this answers is not in the document]
Comment (zimulala) [reply server/server_test.go]: Done.

File: server/server_test.go [hunk 27]
```diff
@@ -537,21 +537,22 @@ func runTestErrorCode(c *C) {
 		_, err = txn2.Exec("select * from tbl_not_exists;")
 		checkErrorCode(c, err, tmysql.ErrNoSuchTable)
 		_, err = txn2.Exec("create database test;")
-		checkErrorCode(c, err, tmysql.ErrDBCreateExists)
+		// Make tests stable. Some times the error may be the ErrInfoSchemaChanged.
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply server/server_test.go]: Get it.

File: ddl/ddl_db_change_test.go [hunk 28]
```diff
@@ -629,3 +631,87 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
...
+
+		se1.SetConnectionID(2)
+		_, err2 := se1.Execute(context.Background(), "alter table test_table add column c2 int")
+		c.Assert(err2, NotNil)
+		c.Assert(strings.Contains(err2.Error(), "Information schema is changed"), IsTrue)
+	}()
+
+	wg.Wait()
+
+	intercept = &ddl.TestInterceptor{}
```
Comment (tiancaiamao) [inline ddl/ddl_db_change_test.go]: This should be set before `go func()`

File: ddl/ddl_db_change_test.go [hunk 29]
```diff
@@ -629,3 +631,87 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
...
+
+		currID := ctx.GetSessionVars().ConnectionID
+		for {
+			seCnt := atomic.LoadInt32(&sessionCnt)
+			// Make sure the two session have got the same information schema. And the first session can continue to go on,
+			// or the frist session finished this SQL(seCnt = finishedCnt), then other sessions can continue to go on.
+			if currID == firstConnID || seCnt == finishedCnt {
+				break
+			}
+			time.Sleep(100 * time.Millisecond)
```
Comment (coocood) [inline ddl/ddl_db_change_test.go]: Why not use `interval`?

File: ddl/ddl_db_change_test.go (hunk 28 above)
Comment (zimulala) [reply ddl/ddl_db_change_test.go]: We set, here we do clean up.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**